### PR TITLE
evals_v2: per-subset nested-LOOC linear-probe training rule (#320)

### LINKED ISSUE #320: probes: library-home the nested-LOCO probe-training + transfer protocol
🤖 *Productionizing the #314 frozen-embedding linear-probe VEP protocol — **training-probes** component. Follows from #314 (kept open as the living research log).*

## Problem

The CV primitives (`traitgym_nested_oof`, `chrom_grouped_oof`, `pair_feature`, `pool_tokens`) are already library-homed and tested in [`variant_probe.py`](https://github.com/Open-Athena/marin-dna/blob/e963ebd/src/marin_dna/pipelines/evals/variant_probe.py). What's still **script-only** is the end-to-end orchestration: loading a bundle + building the feature per the signed/abs rule ([`load_and_pool`/`build_feature`](https://github.com/Open-Athena/marin-dna/blob/e963ebd/scripts/issue314/iter1_representation_sweep.py)), the per-subset nested-LOCO loop ([`iter2_nested.py`](https://github.com/Open-Athena/marin-dna/blob/e963ebd/scripts/issue314/iter2_nested.py)), and the cross-subset/cross-dataset **transfer** loop ([`transfer_oof`, `iter3_transfer.py`](https://github.com/Open-Athena/marin-dna/blob/e963ebd/scripts/issue314/iter3_transfer.py)).

## Scope

1. A tested library entry point: given the pooled `emb_ref`/`emb_alt` columns in the score bundle (#318) + feature combo + protocol config → per-subset OOF scores + selected-`C`s (folds the `iter2_nested` loop into the library). **Simplification:** reading from the in-bundle pooled embeddings, `load_and_pool` collapses to a trivial parquet read + `pair_feature` — no per-token pooling on the production path. **Precision:** upcast the f16 store to **float32 before `pair_feature`** — the `delta`/`abs_delta` subtraction of near-equal allele means is a cancellation risk (see #318's precision note).
2. Encode the **feature rule** (local context + effect): `concat_ref_delta = [ref, alt−ref]` for directional (mendelian, sge), `sum_absdiff = [ref+alt, |alt−ref|]` for swap-invariant (complex, caqtl, dsqtl) — both already in `PAIR_COMBOS` — with a documented `delta`/`abs_delta` (1·D) fallback; mixed-direction transfer uses the common symmetric `sum_absdiff` on both. (Validate `sum_absdiff` > `abs_delta` on caqtl/dsqtl — only confirmed on complex in #314.)
3. `transfer_oof` (cross-subset / cross-dataset nested-`C`) → library + tests.
4. Thin pipeline driver + snakemake rule → per-(model, dataset) probe-metric parquet: probe OOF scored with **#319's per-chrom AUPRC** + global AUPRC + **paired probe-vs-LLR delta** ([`paired_metric_delta_bootstrap`](https://github.com/Open-Athena/marin-dna/blob/e963ebd/src/marin_dna/pipelines/evals/metrics.py)) vs the zero-shot baseline.
5. Keep `load_and_pool` over the per-token cache as a **script-only** escape hatch for re-exploring pooling/innerprod/cov_delta — explicitly out of the production path.

## Dependencies

Depends on **#319** (metric) + **#318** (the in-bundle embeddings); developable now against #314's existing caches.

## Verification

Reproduces #314 iter2/iter3 numbers on the existing models; unit tests on synthetic multi-subset data (coverage, leak-proofing, transfer direction).

### PR DESCRIPTION
**Pull request by gonzalobenegas**
🤖 Productionizes #314's settled frozen-embedding linear-probe VEP protocol — the **probe-training** component (follows #318 inference; precedes the downstream metrics step).

Closes #320.

## What this adds

One **CPU rule** ([`compute_probe`](https://github.com/Open-Athena/marin-dna/blob/ddff494c80fac8c49f725d64450e0a7cf16f8d94/snakemake/analysis/evals_v2/workflow/rules/probe.smk#L22-L69), off `rule all`) that takes the in-bundle pooled embeddings (`emb_ref`/`emb_alt`, #318), trains a per-consequence-subset linear probe, and emits two artifacts per `(model, dataset)`:

- `results/probe/{model}/{dataset}.parquet` — the **leave-one-chromosome-out (LOOC) prediction** for every variant (`probe_score`; `NaN` for skipped subsets), consumed by the downstream metrics issue.
- `results/probe/{model}/{dataset}.joblib` — the **fitted per-subset classifiers** (`{subset: {pipeline, C, feature, n, n_pos, c_summary}}`), serialized for reuse on other datasets.

## The one LOOC protocol (no variants)

- **Feature** ([`pair_feature_from_bundle`](https://github.com/Open-Athena/marin-dna/blob/ddff494c80fac8c49f725d64450e0a7cf16f8d94/src/marin_dna/pipelines/evals/variant_probe.py#L103-L117)) — upcast `emb_ref`/`emb_alt` f16→**f32** before combining (cancellation guard, #318), then `concat_ref_delta` (directional: mendelian/sge) or `sum_absdiff` (swap-invariant: complex/caqtl/dsqtl).
- **Per consequence `subset`**, trained only if `n≥300` **and** `n_chrom≥3`; QTL datasets (no `subset`) collapse to one `all` group; smaller subsets get `NaN` `probe_score` and no classifier.
- **Probe** — fixed `StandardScaler → LogisticRegression(L2)`; the only tuned knob is `C`.
- **CV** ([`traitgym_nested_oof`](https://github.com/Open-Athena/marin-dna/blob/ddff494c80fac8c49f725d64450e0a7cf16f8d94/src/marin_dna/pipelines/evals/variant_probe.py#L140-L185)) — outer leave-one-chromosome-out, inner `GroupKFold` `GridSearchCV` re-tuning `C` per fold; the reusable classifier ([`fit_full_classifier`](https://github.com/Open-Athena/marin-dna/blob/ddff494c80fac8c49f725d64450e0a7cf16f8d94/src/marin_dna/pipelines/evals/variant_probe.py#L187-L230)) is the same pipeline fit on all data. Orchestrated by [`run_subset_probes`](https://github.com/Open-Athena/marin-dna/blob/ddff494c80fac8c49f725d64450e0a7cf16f8d94/src/marin_dna/pipelines/evals/variant_probe.py#L293-L376).

## C grid + verified edge diagnostic

`c_grid = logspace(-12, 4, 17)`, **anchored at both regularization limits** so it can't truncate the optimum — no `inf`, no adaptive widening:
- high `1e4` = a saturation cap (reproduces the unregularized `C→∞` *ranking* bit-for-bit, so `inf` is unnecessary);
- low `1e-12` = a heavy-reg floor (as `C→0` the L2 fit shrinks to the scale-free **mean-difference direction** — *not* a constant predictor, since AUPRC is rank-based).

[`summarize_selected_c`](https://github.com/Open-Athena/marin-dna/blob/ddff494c80fac8c49f725d64450e0a7cf16f8d94/src/marin_dna/pipelines/evals/variant_probe.py#L232-L291) **verifies** every edge pin from the all-data inner-CV curve: `high_edge_gain`/`low_edge_gain` measure whether a pinned edge is still improving vs its interior neighbor, and `truncation_risk` fires only if it is (which the anchored grid avoids). So an edge pin is confirmed *saturated/flat* rather than assumed.

## Verification

24 unit tests; ruff clean; DAG dry-runs to only `compute_probe`. Ran end-to-end on `mix-v0.9-p1B-i24-exp135-m5.1/mendelian_traits` (the one cell already carrying embeddings) on a CPU SkyPilot node:

<details><summary>8-subset parity + edge verification (no truncation anywhere)</summary>

| subset | OOF AUPRC | full_C | edges | low_gain | high_gain | truncation_risk |
|---|---|---|---|---|---|---|
| missense | 0.525 | 1e-3 | clean | −0.0000 | −0.0042 | False |
| splicing | 0.535 | 1e-2 | clean | −0.0000 | −0.0009 | False |
| 3′UTR | 0.499 | 1e-3 | clean | −0.0000 | −0.0029 | False |
| synonymous | 0.498 | **1e4** | 2 high | +0.0000 | +0.0017 | False |
| 5′UTR | 0.483 | 1e-3 | clean | −0.0000 | +0.0005 | False |
| distal | 0.418 | 1e3 | 1 high | −0.0010 | −0.0006 | False |
| ncRNA_exon | 0.319 | 1e-7 | 6 low | +0.0000 | +0.0001 | False |
| tss_proximal | 0.294 | 1e-4 | clean | −0.0000 | −0.0070 | False |
| mature_miRNA | — | _skipped (n=40 < 300)_ | | | | |

All edge gains ≤ 0.007 AUPRC ⟹ the three edge-pinning subsets (synonymous/distal high, ncRNA low) are verified saturated/flat, so `truncation_risk=False` everywhere. OOF AUPRCs track #314 (strong on missense/splicing, weak on ncRNA/tss).
</details>

## Out of scope (handoffs)

- **No metrics** — per-chrom AUPRC (#319), global AUPRC, paired probe-vs-LLR delta. The downstream metrics issue consumes the `probe_score` parquet.
- **No `transfer_oof`** — cross-dataset reuse is via the serialized classifiers, not an explicit transfer loop.
- **Prerequisite:** probing a cell requires its scores parquet to have been produced with `inference.return_embeddings: true`; the rule fails fast otherwise. Only `exp135-m5.1/mendelian` qualifies today — others need a GPU re-score via the #318 overlay first.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: src/marin_dna/pipelines/evals/variant_probe.py [hunk 1]
```diff
@@ -0,0 +1,414 @@
...
+        s = np.asarray(full_c_scores, dtype=float)
+        assert s.shape == grid.shape, (s.shape, grid.shape)
+        high_gain = float(s[-1] - s[-2])  # >tol ⟹ still rising at the weak-reg cap
+        low_gain = float(s[0] - s[1])  # >tol ⟹ wants heavier reg than the floor
+        high_pinned = n_high > 0 or full_c == hi
+        low_pinned = n_low > 0 or full_c == lo
+        out["high_edge_gain"] = high_gain
+        out["low_edge_gain"] = low_gain
+        out["truncation_risk"] = bool(
+            (high_pinned and high_gain > tol) or (low_pinned and low_gain > tol)
```
Comment (gonzalobenegas) [inline src/marin_dna/pipelines/evals/variant_probe.py]: 🤖 **NaN edge-gain silently reports benign** (flagged by 3 finders). If a single-class/failed inner-CV fold leaves `mean_test_score=NaN` at the pinned edge, `high_gain`/`low_gain` is NaN and `nan > tol` is **False** → a real truncation is reported `truncation_risk=False`, defeating the 'verified, not assumed' contract. **Fixed:** a pinned edge with a NaN gain now flags `truncation_risk=True` (can't verify saturation ⇒ don't claim benign). Added `test_summarize_selected_c_nan_edge_gain_flags_risk`.

**Review by gonzalobenegas — COMMENTED** [review]
🤖 **`/code-review max --fix --comment`** — 6 parallel finder angles (correctness × line-by-line/pipeline/cross-file-pitfalls, cleanup, altitude+conventions, tests) + verify. The core prediction path was independently verified sound (positional `oof` assignment, `argsort`/`np.sort` alignment, edge-detection float-equality, gain directionality, `l1_ratio`→L2). Findings are edge-case robustness, defensive-assert gaps, doc/test accuracy, and 2 DRY items — **all fixed in the follow-up commit** except 2 noted-only. Inline comments below; each says how it was addressed.

File: src/marin_dna/pipelines/evals/variant_probe.py [hunk 2]
```diff
@@ -0,0 +1,414 @@
...
+        "full_at_edge": full_at_edge,
+        "at_edge": bool(n_low > 0 or n_high > 0 or full_at_edge),
+        "high_edge_gain": None,
+        "low_edge_gain": None,
+        "truncation_risk": None,
+    }
+    if full_c_scores is not None:
+        s = np.asarray(full_c_scores, dtype=float)
+        assert s.shape == grid.shape, (s.shape, grid.shape)
+        high_gain = float(s[-1] - s[-2])  # >tol ⟹ still rising at the weak-reg cap
```
Comment (gonzalobenegas) [inline src/marin_dna/pipelines/evals/variant_probe.py:292]: 🤖 `s[-2]`/`s[1]` would `IndexError` on a degenerate 1-point `c_grid` (no `len≥2` guard before the gain subtraction). **Fixed:** added `assert grid.size >= 2` in `summarize_selected_c`.

File: src/marin_dna/pipelines/evals/variant_probe.py [hunk 3]
```diff
@@ -0,0 +1,414 @@
...
+    means is a catastrophic-cancellation risk in f16 (see #318), so the upcast
+    must precede ``pair_feature``, not follow it. Returns ``[N, D]`` or ``[N, 2D]``
+    per ``combo``.
+    """
+    assert "emb_ref" in df.columns and "emb_alt" in df.columns, (
+        "scores parquet lacks emb_ref/emb_alt columns — re-score with "
+        "inference.return_embeddings=true (the #318 overlay)"
+    )
+    # f16 store -> f32 BEFORE the difference (cancellation guard).
+    ref = np.stack(df["emb_ref"].to_numpy()).astype(np.float32)
```
Comment (gonzalobenegas) [inline src/marin_dna/pipelines/evals/variant_probe.py:121]: 🤖 `np.stack(df['emb_ref'].to_numpy())` raises an opaque `need at least one array to stack` on an empty frame, before the domain-specific shape assert below can fire. **Fixed:** added `assert len(df) > 0` with a clear message.

File: src/marin_dna/pipelines/evals/variant_probe.py [hunk 4]
```diff
@@ -0,0 +1,414 @@
...
+    alt = np.stack(df["emb_alt"].to_numpy()).astype(np.float32)
+    assert ref.shape == alt.shape and ref.ndim == 2, (ref.shape, alt.shape)
+    assert np.isfinite(ref).all() and np.isfinite(alt).all(), "non-finite embeddings"
+    return pair_feature(ref, alt, combo)
+
+
+def _probe_pipeline() -> Pipeline:
+    """The locked probe: ``StandardScaler → L2 LogisticRegression``.
+
+    ``l1_ratio=0.0`` is the sklearn 1.8 idiom for a pure-L2 penalty (the bare
```
Comment (gonzalobenegas) [inline src/marin_dna/pipelines/evals/variant_probe.py]: 🤖 Comment is factually wrong about sklearn 1.8: it's `penalty` that's mid-deprecation (default resolves to L2), and `l1_ratio` is ignored unless `penalty='elasticnet'` — so `l1_ratio=0.0` is a no-op, not 'the idiom for L2'. The code IS correctly L2, but a maintainer trusting this could 'fix' it to elasticnet and silently change the model. **Fixed:** comment corrected.

File: src/marin_dna/pipelines/evals/variant_probe.py [hunk 5]
```diff
@@ -0,0 +1,414 @@
...
+    y = np.asarray(label).astype(int)
+    g = np.asarray(groups)
+    n = len(y)
+    assert X.ndim == 2 and X.shape[0] == n == len(g), (X.shape, n, len(g))
+    assert len(np.unique(g)) >= 3, "need >=3 groups for nested LOGO"
+    oof = np.full(n, np.nan, dtype=float)
+    selected: list[float] = []
+    for tr, te in LeaveOneGroupOut().split(X, y, g):
+        k = min(inner_splits, len(np.unique(g[tr])))
+        gs = GridSearchCV(
```
Comment (gonzalobenegas) [inline src/marin_dna/pipelines/evals/variant_probe.py]: 🤖 The inner `GridSearchCV(...)` is hand-built identically here and in `fit_full_classifier` — the docstring promises the OOF folds and the shipped classifier tune `C` identically, but a future edit to one spot could silently diverge them. **Fixed:** extracted `_inner_c_search(c_grid, k, n_jobs)`, called from both.

File: src/marin_dna/pipelines/evals/variant_probe.py [hunk 6]
```diff
@@ -0,0 +1,414 @@
...
+    per fold (hence per model and adaptively to its feature dimensionality), with no
+    leakage. Fixed pipeline ``StandardScaler → LogisticRegression`` (L2, no PCA).
+
+    Returns ``(oof_scores, selected_Cs)`` — OOF ``predict_proba[:, 1]`` aligned to
+    ``features`` rows, and the ``C`` chosen on each outer fold (feed to
+    ``summarize_selected_c`` to confirm the grid isn't truncating the optimum).
+    """
+    if c_grid is None:
+        c_grid = DEFAULT_C_GRID
+    X = np.asarray(features, dtype=np.float32)
```
Comment (gonzalobenegas) [inline src/marin_dna/pipelines/evals/variant_probe.py:185]: 🤖 (Noted, not fixed) `traitgym_nested_oof` and `fit_full_classifier` share a 7-line `X/y/g`-coercion preamble. A `_prep_xyg` helper would DRY it, but it's lower-value than the `_inner_c_search` extraction and adds churn — left as-is.

File: src/marin_dna/pipelines/evals/variant_probe.py [hunk 7]
```diff
@@ -0,0 +1,414 @@
...
+    sel = np.asarray(selected_cs, dtype=float)
+    grid = np.sort(np.asarray(c_grid, dtype=float))
+    assert sel.ndim == 1 and sel.size >= 1, sel.shape
+    lo, hi = float(grid[0]), float(grid[-1])
+    n_low = int(np.sum(sel == lo))
+    n_high = int(np.sum(sel == hi))
+    full_at_edge = bool(full_c == lo or full_c == hi)
+    out: dict[str, float | int | bool | None] = {
+        "c_min": float(sel.min()),
+        "c_med": float(np.median(sel)),
```
Comment (gonzalobenegas) [inline src/marin_dna/pipelines/evals/variant_probe.py:278]: 🤖 (Noted, not a bug) `c_med = np.median(sel)` interpolates to a non-grid value for an even fold count (e.g. median of `[1e-2, 1e2]` = 50.005). It's a labelled *median* diagnostic, not 'a selected C', so this is correct-as-named; leaving it.

File: snakemake/analysis/evals_v2/workflow/rules/common.smk [hunk 8]
```diff
@@ -222,3 +224,49 @@ for _d in LL_GAP_CFG.get("datasets", []):
...
+}
+
+
+def get_probe_feature(name):
+    """Probe pair-feature combo for a dataset, from its `score_protocol`; an
+    explicit `probe_feature` on the dataset entry overrides."""
+    cfg = get_dataset_config(name)
+    if "probe_feature" in cfg:
+        return cfg["probe_feature"]
+    return PROBE_FEATURE_BY_PROTOCOL[cfg["score_protocol"]]
```
Comment (gonzalobenegas) [inline snakemake/analysis/evals_v2/workflow/rules/common.smk:263]: 🤖 `PROBE_FEATURE_BY_PROTOCOL[cfg['score_protocol']]` is unguarded — config load only asserts `score_protocol ∈ SCORE_PROTOCOLS` (a *different* dict). A future 3rd protocol added to `SCORE_PROTOCOLS` but not here raises a bare `KeyError` at rule-eval — exactly the late-failure the metrics path is guarded against one screen up. **Fixed:** added a load-time assert that `SCORE_PROTOCOLS ⊆ PROBE_FEATURE_BY_PROTOCOL`.

File: snakemake/analysis/evals_v2/workflow/rules/common.smk [hunk 9]
```diff
@@ -222,3 +224,49 @@ for _d in LL_GAP_CFG.get("datasets", []):
...
+
+
+# Fail fast: every probe entry names a known checkpoint and lists datasets that
+# model is actually evaluated on; any explicit `probe_feature` override is a known
+# combo. A typo would otherwise surface late inside the rule.
+for _pm in PROBE_MODELS:
+    assert _pm["name"] in MODELS, f"probe model {_pm['name']!r} not found in `models`"
```
Comment (gonzalobenegas) [inline snakemake/analysis/evals_v2/workflow/rules/common.smk:274]: 🤖 `_pm['name']`/`_pm['datasets']` are accessed directly, so a probe entry written as a bare string (mirroring the sibling `models:` of calibration/umap) raises `TypeError: string indices must be integers` instead of this fail-fast block's intended message. **Fixed:** validate each entry is a mapping with `name` + `datasets` first (this also protects the `rule probe` collector, which runs after config load).

File: snakemake/analysis/evals_v2/workflow/rules/probe.smk [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+lives in `marin_dna.pipelines.evals.variant_probe.run_subset_probes`; this rule is thin
+glue.
+
+CPU-only (sklearn on cached embeddings — no GPU). Requires the input scores parquet to
+carry `emb_ref`/`emb_alt`, i.e. it must have been produced with
+`inference.return_embeddings: true` (the #318 overlay); the rule fails fast otherwise.
+
+Kept OFF `rule all` (a few-models analysis, like `calibration` / `umap`); build by name:
+
+    snakemake probe                                   # every configured probe cell
```
Comment (gonzalobenegas) [inline snakemake/analysis/evals_v2/workflow/rules/probe.smk]: 🤖 The documented `snakemake probe` commands omit `--rerun-triggers mtime`. The scores parquet was built with the #318 overlay (`return_embeddings: true`) but the committed default is `false`, so the default `params` trigger can schedule a `compute_scores` rerun — which on a CPU probe node (no GPU/gcloud) regenerates the parquet **without** `emb_ref`/`emb_alt`, then the rule's assert fails. **Fixed:** the rule docstring + README build commands now include `--rerun-triggers mtime` with the rationale.

File: tests/pipelines/evals/test_variant_probe.py [hunk 11]
```diff
@@ -0,0 +1,362 @@
...
+        "emb_ref": list(ref.astype(np.float16)),
+        "emb_alt": list(alt.astype(np.float16)),
+    }
+    data.update(cols)
+    return pd.DataFrame(data)
+
+
+def test_pair_feature_from_bundle_upcasts_f16_before_combining() -> None:
+    # values exact in f16 so the only thing under test is the upcast + combine.
+    ref = np.array([[1.0, 2.0], [0.5, 4.0]], dtype=np.float32)
```
Comment (gonzalobenegas) [inline tests/pipelines/evals/test_variant_probe.py]: 🤖 This 'upcasts f16 before combining' test uses f16-**exact** values with `delta`, so it can't distinguish upcast-before vs after-subtract (f16 subtraction of near-equal values is exact by Sterbenz) — its stated purpose. The real guard is `matches_manual_stack` below. **Fixed:** rewrote it to use `sum_absdiff` with `1500/1501` (exact in f16, but their sum `3001` rounds to `3000` in f16) so a missing upcast returns f16 `[[3000,1]]` and fails.

File: tests/pipelines/evals/test_variant_probe.py [hunk 12]
```diff
@@ -0,0 +1,362 @@
...
+        ref = rng.standard_normal((n, d)).astype(np.float32)
+        delta = (rng.standard_normal((n, d)) * 0.1).astype(np.float32)
+        delta[:, 0] += signal * label
+        frames.append(
+            _emb_frame(ref, ref + delta, chrom=chrom, label=label, subset=name)
+        )
+    return pd.concat(frames, ignore_index=True)
+
+
+def test_run_subset_probes_covers_trained_subsets_and_drops_emb() -> None:
```
Comment (gonzalobenegas) [inline tests/pipelines/evals/test_variant_probe.py:303]: 🤖 Coverage gaps: the single-class skip (`0 < n_pos < n`), the `min_chroms` skip disjunct, the per-subset `try/except` resilience, and the **low-edge** `truncation_risk=True` branch were all untested (only the high-edge truncation was). **Fixed:** added `test_run_subset_probes_skips_single_class_subset`, `_skips_below_min_chroms`, `_keeps_going_when_one_subset_errors` (monkeypatched), and `test_summarize_selected_c_low_edge_truncation_flagged`.

**Comment by gonzalobenegas** [comment]
🤖 Fixes for the review above are in `7bcb24a` — 11 findings addressed (NaN-edge truncation flag, grid/empty-frame guards, corrected `l1_ratio` comment, `_inner_c_search` dedup, the two `common.smk` fail-fast guards, the `--rerun-triggers mtime` docs, the real f16-upcast test, and 5 coverage tests; **24 → 29 tests pass**, ruff + dry-run clean). Two left as noted-only: the `_prep_xyg` DRY (low value) and `c_med` (a correctly-named median). One **deferred** finding not in this PR's diff: `joblib` is imported by first-party code (`common.smk`/`probe.smk`) but only a transitive sklearn dep — worth a one-line `pyproject` add (needs a lockfile update), best as a separate change.
